thorlcr/thorutil/thbuf.cpp

```diff
@@ -1658,6 +1661,445 @@ ISharedSmartBuffer *createSharedSmartMemBuffer(CActivityBase *activity, unsigned
 }
 
 
+class CSharedFullSpillingWriteAhead : public CInterfaceOf<ISharedRowStreamReader>
+{
+    typedef std::vector<const void *> Rows;
+    class COutputRowStream : public CSimpleInterfaceOf<IRowStream>
+    {
+        CSharedFullSpillingWriteAhead &owner;
+        unsigned whichOutput = 0;
+        size32_t localRowsIndex = 0;
+        rowcount_t currentRow = 0;
+        Rows rows;
+        Owned<IBufferedSerialInputStream> inputStream;
+        CThorStreamDeserializerSource ds;
+        std::atomic<bool> eof = false;
+
+        inline const void *getClearRow(unsigned i)
+        {
+            const void *row = rows[i];
+            rows[i] = nullptr;
+            return row;
+        }
+        void freeRows()
+        {
+            for (auto it = rows.begin() + localRowsIndex; it != rows.end(); ++it)
+                ReleaseThorRow(*it);
+            rows.clear();
+            localRowsIndex = 0;
+        }
+        const void *getRowFromStream()
+        {
+            if (currentRow == lastKnownWritten)
+            {
+                if (!owner.checkWriteAhead(lastKnownWritten))
+                {
+                    eof = true;
+                    return nullptr;
+                }
+            }
+            if (owner.inputGrouped)
+            {
+                bool eog;
+                inputStream->read(sizeof(bool), &eog);
+                if (eog)
+                {
+                    currentRow++;
+                    return nullptr;
+                }
+            }
+            currentRow++;
+            RtlDynamicRowBuilder rowBuilder(owner.allocator);
+            size32_t sz = owner.deserializer->deserialize(rowBuilder, ds);
+            return rowBuilder.finalizeRowClear(sz);
+        }
+    public:
+        rowcount_t lastKnownWritten = 0;
```

Comment:
lastKnownAvailable would be a better name to cover both in-memory and spilled rows.

Reply:
agree, renamed.